pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,646 @@
+			ownerRefs := []metav1.OwnerReference{
+				*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+			}
+			if err := kutil.CreateOrUpdateSecretByReference(ctx, a.gardenClient.Client(), &spec.Backup.SecretRef, corev1.SecretTypeOpaque, shootSecret.Data, ownerRefs); err != nil {
+				return err
+			}
+		}
+	}
+
+	// If secret reference is specified, create or update the corresponding secret
+	if spec.SecretRef != nil {
+		// Get shoot kubeconfig secret
+		shootKubeconfigSecret, err := a.getShootKubeconfigSecret(ctx, shoot)
+		if err != nil {
+			return err
+		}
+
+		// Initialize seed secret data
+		data := shootSecret.Data
+		data[kubernetes.KubeConfig] = shootKubeconfigSecret.Data[kubernetes.KubeConfig]
+
+		// Create or update seed secret
+		ownerRefs := []metav1.OwnerReference{
+			*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+		}
+		if err := kutil.CreateOrUpdateSecretByReference(ctx, a.gardenClient.Client(), spec.SecretRef, corev1.SecretTypeOpaque, data, ownerRefs); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *actuator) unprepareSeedSpec(ctx context.Context, spec *gardencorev1beta1.SeedSpec, managedSeed *seedmanagementv1alpha1.ManagedSeed) error {
+	// If backup is specified, delete the backup secret if it exists and is owned by the managed seed
+	if spec.Backup != nil {
+		// Get backup secret
+		backupSecret, err := kutil.GetSecretByReference(ctx, a.gardenClient.Client(), &spec.Backup.SecretRef)
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+
+		// Delete backup secret if it exists and is owned by the managed seed
+		if !apierrors.IsNotFound(err) && metav1.IsControlledBy(backupSecret, managedSeed) {
+			if err := kutil.DeleteSecretByReference(ctx, a.gardenClient.Client(), &spec.Backup.SecretRef); err != nil {
+				return err
+			}
+		}
+	}
+
+	// If secret reference is specified, delete the corresponding secret
+	if spec.SecretRef != nil {
+		if err := kutil.DeleteSecretByReference(ctx, a.gardenClient.Client(), spec.SecretRef); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *actuator) getShootSecret(ctx context.Context, shoot *gardencorev1beta1.Shoot) (*corev1.Secret, error) {
+	shootSecretBinding := &gardencorev1beta1.SecretBinding{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, shoot.Spec.SecretBindingName), shootSecretBinding); err != nil {
+		return nil, err
+	}
+	return kutil.GetSecretByReference(ctx, a.gardenClient.Client(), &shootSecretBinding.SecretRef)
+}
+
+func (a *actuator) getShootKubeconfigSecret(ctx context.Context, shoot *gardencorev1beta1.Shoot) (*corev1.Secret, error) {
+	shootKubeconfigSecret := &corev1.Secret{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, fmt.Sprintf("%s.kubeconfig", shoot.Name)), shootKubeconfigSecret); err != nil {
```

Comment:
Can we expose https://github.com/gardener/gardener/blob/500bfa8d3466ff37633291584e1efd992f67351a/pkg/operation/botanist/secrets.go#L278 (and the others) as a constant and use it here? Maybe `pkg/operation/common/types.go` is a good place for now?

Reply:
OK, perhaps in a new PR, since this is rather an unrelated change?

Reply:
Sure, no strong opinion.